src/functions/PiecewiseMultilinearFromReporter.C

```diff
@@ -0,0 +1,137 @@
+
+registerMooseObject("isopodApp", PiecewiseMultilinearFromReporter);
+
+InputParameters
+PiecewiseMultilinearFromReporter::validParams()
+{
+  InputParameters params = PiecewiseMultiInterpolationFromReporter::validParams();
+  params.addClassDescription(
+      "This is a copy of PiecewiseMultilinear which performs linear interpolation on 1D, 2D, 3D or "
+      "4D "
+      "data, except it is derived from PiecewiseMultiInterpolationFromReporter. If a point lies "
+      "outside the data range, the appropriate end "
+      "value is used.");
+  params.addParam<Real>(
+      "epsilon", 1e-12, "Finite differencing parameter for gradient and time derivative");
+  return params;
+}
+
+PiecewiseMultilinearFromReporter::PiecewiseMultilinearFromReporter(
+    const InputParameters & parameters)
+  : PiecewiseMultiInterpolationFromReporter(parameters), _epsilon(getParam<Real>("epsilon"))
+{
+}
+
+Real
+PiecewiseMultilinearFromReporter::sample(const GridPoint & pt) const
+{
+  return sampleInternal<false>(pt);
+}
+
+ADReal
+PiecewiseMultilinearFromReporter::sample(const ADGridPoint & pt) const
+{
+  return sampleInternal<true>(pt);
+}
+
+template <bool is_ad>
+MooseADWrapper<Real, is_ad>
+PiecewiseMultilinearFromReporter::sampleInternal(const MooseADWrapper<GridPoint, is_ad> pt) const
+{
+  /*
+   * left contains the indices of the point to the 'left', 'down', etc, of pt
+   * right contains the indices of the point to the 'right', 'up', etc, of pt
+   * Hence, left and right define the vertices of the hypercube containing pt
+   */
+  GridIndex left(_dim);
+  GridIndex right(_dim);
+  for (unsigned int i = 0; i < _dim; ++i)
+    getNeighborIndices(_grid[i], MetaPhysicL::raw_value(pt[i]), left[i], right[i]);
+
+  /*
+   * The following just loops through all the vertices of the
+   * hypercube containing pt, evaluating the function at all
+   * those vertices, and weighting the contributions to the
+   * final result depending on the distance of pt from the vertex
+   */
+  MooseADWrapper<Real, is_ad> f = 0;
+  MooseADWrapper<Real, is_ad> weight;
+  GridIndex arg(_dim);
+  // number of points in hypercube = 2^_dim
+  for (unsigned int i = 0; i < (1u << _dim); ++i)
+  {
+    weight = 1;
+    for (unsigned int j = 0; j < _dim; ++j)
+      // shift i j-bits to the right and see if the result has a 0 as its right-most bit
+      if ((i >> j) % 2 == 0)
+      {
+        arg[j] = left[j];
+        if (left[j] != right[j])
+          weight *= std::abs(pt[j] - _grid[j][right[j]]);
+        else
+          // unusual "end condition" case. weight by 0.5 because we will encounter this twice
+          weight *= 0.5;
+      }
+      else
+      {
+        arg[j] = right[j];
+        if (left[j] != right[j])
+          weight *= std::abs(pt[j] - _grid[j][left[j]]);
+        else
+          // unusual "end condition" case. weight by 0.5 because we will encounter this twice
+          weight *= 0.5;
+      }
+    f += _gridded_data->evaluateFcn(arg) * weight;
```

Comment:
Can you read this from a reporter instead of from gridded data? That will allow us to change the parameter values without having to reset the subapp.

Reply:
Don't do this change, it would require a new public function on gridded data that would give you the index.